CurrencyConverterControl/CurrencyConverterControl.cs

```diff
@@ -0,0 +1,251 @@
+        /// It is bound to the user interface elements where the converted amount is displayed to the user.
+        /// </remarks>
+         [Description("Converted value")]
+        public double OutputValue
+        {
+            get { return (double)GetValue(OutputValueProperty); }
+            set { SetValue(OutputValueProperty, value); }
+        }
+
+        public static readonly DependencyProperty OutputValueProperty =
+           DependencyProperty.Register("OutputValue", typeof(double), typeof(CurrencyConverterControl));
+        
+        public static readonly DependencyProperty DestinationCurrencyProperty =
+           DependencyProperty.Register("DestinationCurrency", typeof(Currency), typeof(CurrencyConverterControl));
+        /// <summary>
+        /// Gets the Destination Currency to perform the currency conversion  
+        /// </summary>
+        [Description("Value of the destination currency")]
+        public Currency DestinationCurrency
+        {
+            get { return (Currency)GetValue(DestinationCurrencyProperty); }
+            set { SetValue(DestinationCurrencyProperty, value); }
+        }
+
+        public static readonly DependencyProperty SourceCurrencyProperty =
+            DependencyProperty.Register("SourceCurrency", typeof(Currency), typeof(CurrencyConverterControl));
+
+        /// <summary>
+        /// Gets the Source Currency to perform the currency conversion  
+        /// </summary>
+         [Description("Value of the source currency")]
+        public Currency SourceCurrency
+        {
+            get { return (Currency)GetValue(SourceCurrencyProperty); }
+            set { SetValue(SourceCurrencyProperty, value); }
+        }
+
+        /// <summary>
+        /// Gets or sets the ObservableCollection of Currency objects.
+        /// </summary>
+        /// <remarks>
+        /// The CurrencyList property represents a collection of Currency objects,
+        /// typically used for storing a list of currencies available in the application
+        /// to fill in the ComboBox element.
+        /// It is initialized with a new instance of ObservableCollection<Currency>.
+        /// </remarks>
+        [Description("Collection of currencies list")]
+        public ObservableCollection<Currency> CurrencyList { get; private set; } = new ObservableCollection<Currency>();
+
+        /// <summary>
+        /// Property to allow replacement of the currency converter
+        /// </summary>
+        public ICurrencyDataProvider CurrencyConverter
+        {
+            get { return CurrencyDataProvider; }
+            set { CurrencyDataProvider = value ?? throw new ArgumentNullException(nameof(value)); }
+        }
+        #endregion
+        public override void OnApplyTemplate()
+        {
+           
+            LoadCurrencies();
+            _currencyDestination = Template.FindName("CmbDestination", this) as ComboBox;
```

Comment:
I already wrote about it, but maybe my comment got lost. Note, people can make own control templates for our controls. If someone set different name for ComboBox, or don't include ComboBox at all, below code will fail with NullReferenceException. 
You should check whether combobox is null here and fire own exception with some reasonable text like "The required template part CmbDestination is missing"

Reply:
oki l am adding a check if OnApplyTemplate are found or not this two elements and if not throw the exception. 
thanks